iceoryx_examples/ara_com/ara/runtime.hpp

```diff
@@ -0,0 +1,171 @@
+        {
+            araServiceContainer.push_back({serviceIdentifier, instanceIdentifier});
+        }
+
+        return araServiceContainer;
+    }
+
+    ara::com::FindServiceHandle StartFindService(ara::com::FindServiceHandler<com::ProxyHandleType> handler,
+                                                 ara::com::ServiceIdentifier& serviceIdentifier,
+                                                 ara::com::InstanceIdentifier& instanceIdentifier) noexcept
+    {
+        /// @todo #1332 Are duplicate entries allowed?
+        m_callbacks.push_back({handler, {serviceIdentifier, instanceIdentifier}});
+
+        if (m_callbacks.size() == 1)
+        {
+            auto invoker = iox::popo::createNotificationCallback(invokeCallback, *this);
+            m_listener.attachEvent(m_discovery, iox::runtime::ServiceDiscoveryEvent::SERVICE_REGISTRY_CHANGED, invoker)
+                .or_else([](auto) {
+                    std::cerr << "unable to attach discovery" << std::endl;
+                    std::exit(EXIT_FAILURE);
+                });
+        }
+
+        return ara::com::FindServiceHandle({serviceIdentifier, instanceIdentifier});
+    }
+
+    void StopFindService(ara::com::FindServiceHandle handle) noexcept
+    {
+        auto iter = m_callbacks.begin();
+        for (; iter != m_callbacks.end(); iter++)
+        {
+            if (iter->second.GetServiceId() == handle.GetServiceId()
+                && iter->second.GetInstanceId() == handle.GetInstanceId())
+            {
+                break;
+            }
+        }
+        if (iter != m_callbacks.end())
+        {
+            m_callbacks.erase(iter);
+        }
+
+        if (m_callbacks.empty())
+        {
+            m_listener.detachEvent(m_discovery, iox::runtime::ServiceDiscoveryEvent::SERVICE_REGISTRY_CHANGED);
+        }
+    }
+
+  private:
+    explicit Runtime() noexcept = default;
+
+    bool verifyThatServiceIsComplete(ara::com::ServiceHandleContainer<com::ProxyHandleType>& container)
+    {
+        // The service level of AUTOSAR Adaptive is not available in iceoryx, instead every publisher and server is
+        // considered as a service. A ara::com binding implementer would typically query the AUTOSAR meta model here, to
+        // find out if all event, fields and methods of a service are available. For the example we assume that the
+        // 'MinimalSkeleton' service is complete when the container contains the three iceoryx services:
+        //
+        // 1. EventPublisher: MinimalSkeleton, Instance, Event
+        // 2. FieldPublisher: MinimalSkeleton, Instance, Field
+        // 3. MethodServer:   MinimalSkeleton, Instance, Method
+
+        if (container.size() == 3U)
+        {
+            return true;
+        }
+        return false;
+    }
+
+    static void invokeCallback(iox::runtime::ServiceDiscovery*, Runtime* self)
+    {
+        // 1) Has the availability of one of the registered services changed?
```

Comment:
More precisely: if the service discovery changed at all (can be due to another service we did not register here), this gets invoked. Then we iterate over the elements of the container which are, not quite appropriately, called callbacks but are pairs of callbacks and identifiers. This needs clarification/changes because is not intuitive.

We then try to find the service instance associated with the callback, and get back a container of instances (which could be empty).

Now we call the callback all the time with this container, but if I remember correctly this is wrong as ara demands only to call it if the availability has changed. We can only track this with extra state for each/service instance pair where we track whether it was available last or not (a bool). I recommend abstracting the service state and managing it in the array which is currently called `m_callbacks`.

I will look into the spec tomorrow to verify whether my understanding of Autosar is correct.

Assuming it is, we can almost correctly implement it. We may miss a rapid switch between available and unavailable and back to available (basically ABA problem) and not trigger the callback for this service (which should be called twice). We may have been woken up by the listener due to some other service change. The problem is that we only ever know the last state of discovery. This is not easily fix and generally not a big deal for usefulness. I am fairly sure the currently implemented behavior is not what the Autosar spec requires.

TLDR: the tracking of available services needs major rework to (almost) comply with Autosar IMO. It is not possible to perfectly comply with Autosar due to the way our discovery works. To perfectly comply with Autosar each instance we track needs its own semaphore/condition variable (stored in shared memory) and this would be fairly expensive.

Reply:
We can discuss this with @budrus if this workaround can stay for simplicity or if we should fix this. Anyway, I added a comment explaining why we call the callback on ANY change.

Reply:
I verified that the callback is only to be called if the availability of the service in question changes. See https://github.com/eclipse-iceoryx/iceoryx/pull/1343#issuecomment-1114595469